apps/trading/client-pages/swap/swap-container.tsx

```diff
@@ -0,0 +1,307 @@
+        const quoteAsset = getQuoteAsset(m);
+        if (baseAsset && assetsData) {
+          spotAssets[baseAsset.id] = assetsData[baseAsset.id];
+        }
+        if (quoteAsset && assetsData) {
+          spotAssets[quoteAsset.id] = assetsData[quoteAsset.id];
+        }
+        return true;
+      }
+      return false;
+    });
+    return { spotMarkets, spotAssets };
+  }, [assetsData, markets]);
+
+  const chooseMarket = useCallback(() => {
+    if (!spotMarkets || !baseAsset || !quoteAsset) return;
+    return spotMarkets.find((m) => {
+      const mBaseAsset = getBaseAsset(m);
+      const mQuoteAsset = getQuoteAsset(m);
+      if (!mBaseAsset || !mQuoteAsset) return false;
+      if (mBaseAsset.id === baseAsset.id && mQuoteAsset.id === quoteAsset.id) {
+        setSide(Side.SIDE_BUY);
+        return true;
+      }
+      if (mBaseAsset.id === quoteAsset.id && mQuoteAsset.id === baseAsset.id) {
+        setSide(Side.SIDE_SELL);
+        return true;
+      }
+      return false;
+    });
+  }, [baseAsset, quoteAsset, spotMarkets]);
+
+  const accountAssetIds = accounts?.map((a) => a.asset.id);
+
+  const quoteAssetBalance = useMemo(
+    () => assetBalance(quoteAsset, accounts),
+    [accounts, quoteAsset]
+  );
+  const baseAssetBalance = useMemo(
+    () => assetBalance(baseAsset, accounts),
+    [accounts, baseAsset]
+  );
+
+  const marketPrice = useMemo(() => {
+    return (
+      (side === Side.SIDE_BUY
+        ? marketData?.bestOfferPrice // best ask
+        : marketData?.bestBidPrice) || // best bid
+      marketData?.lastTradedPrice
+    );
+  }, [marketData, side]);
+
+  const orderSubmission = useMemo(() => {
+    if (!market || !side || !quoteAmount || quoteAmount === '0') return;
+
+    const toleranceFactor = priceImpactTolerance
+      ? Number(priceImpactTolerance) / 100
+      : 0;
+
+    const price = derivePrice(marketData, side, toleranceFactor, market);
+
+    return {
+      marketId,
+      side,
+      type: price ? OrderType.TYPE_LIMIT : OrderType.TYPE_MARKET,
+      price: price ? price.toFixed(0) : undefined,
+      timeInForce: OrderTimeInForce.TIME_IN_FORCE_FOK,
+      size: removeDecimal(quoteAmount, market.positionDecimalPlaces),
+    };
+  }, [market, marketData, marketId, priceImpactTolerance, quoteAmount, side]);
+
+  useEffect(() => {
+    const market = chooseMarket();
+    setMarketId(market?.id || '');
+    setMarket(market);
+  }, [chooseMarket, setMarketId, setMarket]);
+
+  const switchAssets = () => {
+    const newBaseAsset = quoteAsset;
+    const newQuoteAsset = baseAsset;
+    setBaseAsset(newBaseAsset);
+    setQuoteAsset(newQuoteAsset);
+    newBaseAsset && newQuoteAsset && chooseMarket();
+  };
+
+  const switchAmounts = () => {
+    setValue('quoteAmount', baseAmount);
+  };
+
+  const onSubmit = () => {
+    if (!marketId || !side || !market || !orderSubmission) return;
+    create({ orderSubmission });
+  };
+
+  useEffect(() => {
+    if (!orderSubmission || !market) return;
```

Comment:
Setting state from within an effect like this is a code smell. Its extremely hard to know how and when this should get triggered. 

It looks like you are setting a price value based on a number of different inputs. I would much rather:

- just derive the value based on other state and/or
- calculate it once on the user clicking submit

Reply:
I went for the first option. Could you please review again?